File: calculations/watch.py
"""
 *****************************************************************************
   FILE:        watch.py

   AUTHOR: Nathan Koike

   ASSIGNMENT: Project 2

   DATE: 9/6/17

   DESCRIPTION: make a program that tells the correct time of an upside down
                watch

 *****************************************************************************
"""


def main():
    """ takes user inputted times and converts them to be correct on an
        upside down watch"""
    # RAW is an unprocessed time
    timeRAW = str(input("What time does your upside-down watch read (hours:minutes)? "))
    # conversions to the corrected time
    timeHrs = int(timeRAW[0:(timeRAW.index(':'))])
    timeMins = int(timeRAW[(timeRAW.index(':'))+1:])
    timeHrs_inMins = 60 * timeHrs
    timeMins_total = timeHrs_inMins + timeMins
    timeMins_fixed = timeMins_total + 390
    # final here refers to the corrected time not the time that will print out
    timeHrs_final = int(timeMins_fixed // 60)
    timeMins_final = int(timeMins_fixed % 60)

    # disp denotes a time that will print
    # this if/else statement exists to return hours in a 12-hour format and can
    # be taken out if a 24-hour format is desired
    if timeHrs_final > 12:
        timeHrs_disp = str(timeHrs_final - 12)
    else:
        timeHrs_disp = str(timeHrs_final)

    timeMins_dispBase = str(timeMins_final)
    zero_timeMins_dispBase = '0'+timeMins_dispBase
    timeMins_disp = zero_timeMins_dispBase[-2:]

    print("The right-side-up time is:", timeHrs_disp + ':' + timeMins_disp)


# this invokes the main function.  It is always included in our
# python programs
if __name__ == "__main__":
    main()

File: calculations/watch.js
const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
  crlfDelay: Infinity
});

function locationsOfLetter(letter, string) {
  let locations = [];

  for (let i = 0; i < string.length; i++)
    if (string[i] == letter) locations.push(i);

  return locations;
}

function locationsOfSubstr(substring, string) {
  let locations = [];

  for (let i = 0; i < string.length; i++)
    if (string.substr(i, substring.length) == substring) locations.push(i);

  return locations;
}

function divideString(string, positions) {
  let subs = [];
  let substring = "";

  for (let i = 0; i < string.length; i++) {
    if (i == positions[0]) {
      positions.shift();
      subs.push(substring);
      substring = "";
    }
    substring += string[i];
  }

  subs.push(substring);

  return subs;
}

function split(string, separator) {
  let ans = divideString(string, locationsOfSubstr(separator, string));

  for (let i = 1; i < ans.length; i++)
    ans[i] = ans[i].substr(separator.length, ans[i].length - separator.length);

  return ans;
}

function main() {
  readline.question(
    "What time does your upside-down watch read (hours:minutes)? ",
    time => {
      let [timeHrs, timeMin] = split(time, ":").map(x => parseInt(x));

      // 390: 6 hours + 30 mins
      let fixedMins = timeHrs * 60 + timeMin + 390;

      let fixedHours = Math.floor(fixedMins / 60);
      fixedMins -= fixedHours * 60;

      if (fixedHours > 12) fixedHours -= 12;

      console.log(
        "The right-side-up time is:",
        (fixedHours.length < 2 ? "0" + fixedHours : fixedHours) +
          ":" +
          (fixedMins.length < 2 ? "0" + fixedMins : fixedMins)
      );
    }
  );
}

main();
